src/sdiv.rs

```diff
@@ -0,0 +1,121 @@
+use Int;
+
+macro_rules! div {
+    ($intrinsic:ident: $ty:ty, $uty:ty) => {
+        /// Returns `a / b`
+        #[cfg_attr(not(test), no_mangle)]
+        pub extern "C" fn $intrinsic(a: $ty, b: $ty) -> $ty {
+            let s_a = a >> (<$ty>::bits() - 1);
+            let s_b = b >> (<$ty>::bits() - 1);
+            let a = (a ^ s_a) - s_a;
+            let b = (b ^ s_b) - s_b;
+            let s = s_a ^ s_b;
+            let r = (a as $uty) / (b as $uty);
+            (r as $ty ^ s) - s
+        }
+    }
+}
+
+macro_rules! mod_ {
+    ($intrinsic:ident: $ty:ty, $uty:ty) => {
+        /// Returns `a % b`
+        #[cfg_attr(not(test), no_mangle)]
+        pub extern "C" fn $intrinsic(a: $ty, b: $ty) -> $ty {
+            let s = b >> (<$ty>::bits() - 1);
+            let b = (b ^ s) - s;
+            let s = a >> (<$ty>::bits() - 1);
+            let a = (a ^ s) - s;
+            let r = (a as $uty) % (b as $uty);
+            (r as $ty ^ s) - s
+        }
+    }
+}
+
+macro_rules! divmod {
+    ($intrinsic:ident, $div:ident: $ty:ty, $uty:ty) => {
+        /// Returns `a / b` and sets `*rem = n % d`
+        #[cfg_attr(not(test), no_mangle)]
+        pub extern "C" fn $intrinsic(a: $ty, b: $ty, rem: &mut $ty) -> $ty {
+            let r = $div(a, b);
+            *rem = a - (r * b);
+            r
+        }
+    }
+}
+
+div!(__divsi3: i32, u32);
+div!(__divdi3: i64, u64);
+mod_!(__modsi3: i32, u32);
+mod_!(__moddi3: i64, u64);
+divmod!(__divmodsi4, __divsi3: i32, u32);
+divmod!(__divmoddi4, __divdi3: i64, u64);
```

Comment:
It's only called by the ARM builtins indirectly. I'll add it to the README anyways.